File: rp2350-reference/split-datasheet.clj
(ns split-datasheet
  (:require [clojure.java.io :as io]
            [clojure.java.shell :as shell]
            [clojure.string :as str]))

(def printed-ranges
  [{:name "01-introduction" :start 13 :end 23}
   {:name "02-system-bus" :start 24 :end 34}
   {:name "03a-processor-SIO" :start 35 :end 81}
   {:name "03b-processor-interrupts-debug" :start 82 :end 99}
   {:name "03c-cortex-m33-coprocessors" :start 100 :end 122}
   {:name "03d-cortex-m33-processor" :start 123 :end 232}
   {:name "03e-hazard3-processor" :start 233 :end 334}
   {:name "03f-architecture-switching" :start 335 :end 336}
   {:name "04-memory" :start 337 :end 352}
   {:name "05a-bootrom-concepts" :start 353 :end 374}
   {:name "05b-bootrom-apis" :start 375 :end 415}
   {:name "05c-bootrom-usb-uart" :start 416 :end 440}
   {:name "06-power" :start 441 :end 493}
   {:name "07-resets" :start 494 :end 512}
   {:name "08a-clocks-overview" :start 513 :end 553}
   {:name "08b-clocks-oscillators" :start 554 :end 586}
   {:name "09a-gpio-overview" :start 587 :end 603}
   {:name "09b-gpio-io-user-bank" :start 604 :end 759}
   {:name "09c-gpio-io-qspi-pads" :start 760 :end 815}
   {:name "10-security" :start 816 :end 875}
   {:name "11a-pio-overview-model" :start 876 :end 888}
   {:name "11b-pio-instructions" :start 889 :end 901}
   {:name "11c-pio-details-examples" :start 902 :end 960}
   {:name "12a-uart" :start 961 :end 982}
   {:name "12b-i2c" :start 983 :end 1045}
   {:name "12c-spi" :start 1046 :end 1065}
   {:name "12d-adc-temp" :start 1066 :end 1075}
   {:name "12e-pwm" :start 1076 :end 1093}
   {:name "12f-dma" :start 1094 :end 1140}
   {:name "12g-usb" :start 1141 :end 1181}
   {:name "12h-timers-watchdog" :start 1182 :end 1201}
   {:name "12i-hstx" :start 1202 :end 1211}
   {:name "12j-trng-sha256" :start 1212 :end 1225}
   {:name "12k-qspi-qmi" :start 1226 :end 1248}
   {:name "12l-system-control" :start 1249 :end 1267}
   {:name "13-otp" :start 1268 :end 1326}
   {:name "14-electrical" :start 1327 :end 1348}
   {:name "15-appendices" :start 1349 :end 1378}])

(def bookmark-checks
  [{:range-name "01-introduction" :title-prefix "Chapter 1. "}
   {:range-name "02-system-bus" :title-prefix "Chapter 2. "}
   {:range-name "03a-processor-SIO" :title-prefix "Chapter 3. "}
   {:range-name "04-memory" :title-prefix "Chapter 4. "}
   {:range-name "05a-bootrom-concepts" :title-prefix "Chapter 5. "}
   {:range-name "06-power" :title-prefix "Chapter 6. "}
   {:range-name "07-resets" :title-prefix "Chapter 7. "}
   {:range-name "08a-clocks-overview" :title-prefix "Chapter 8. "}
   {:range-name "09a-gpio-overview" :title-prefix "Chapter 9. "}
   {:range-name "10-security" :title-prefix "Chapter 10. "}
   {:range-name "11a-pio-overview-model" :title-prefix "Chapter 11. "}
   {:range-name "12a-uart" :title-prefix "Chapter 12. "}
   {:range-name "13-otp" :title-prefix "Chapter 13. "}
   {:range-name "14-electrical" :title-prefix "Chapter 14. "}
   {:range-name "15-appendices" :title-prefix "Appendix A:"}])

(defn die [message]
  (binding [*out* *err*]
    (println (str "error: " message)))
  (System/exit 1))

(defn warn [message]
  (binding [*out* *err*]
    (println (str "warning: " message))))

(defn run-capture [command & args]
  (let [{:keys [exit out err]} (apply shell/sh command args)]
    (if (zero? exit)
      out
      (die (str "command failed: "
                (str/join " " (cons command args))
                " ("
                (if (str/blank? err) (str "exit " exit) (str/trim err))
                ")")))))

(defn run-command [command & args]
  (let [{:keys [exit err]} (apply shell/sh command args)]
    (when-not (zero? exit)
      (die (str "command failed: "
                (str/join " " (cons command args))
                " ("
                (if (str/blank? err) (str "exit " exit) (str/trim err))
                ")")))))

(defn require-tool [tool-name]
  (let [{:keys [exit]} (shell/sh "sh" "-lc" (str "command -v " tool-name))]
    (when-not (zero? exit)
      (die (str tool-name " is required but not installed")))))

(defn parse-int [text context]
  (try
    (Integer/parseInt (str/trim text))
    (catch Exception _
      (die (str "invalid integer for " context ": " text)))))

(defn parse-bookmarks [bookmark-dump]
  (loop [lines (str/split-lines bookmark-dump)
         current-title nil
         bookmarks []]
    (if (empty? lines)
      (if (empty? bookmarks)
        (die "no bookmarks found in PDF metadata")
        bookmarks)
      (let [line (str/trim (first lines))]
        (cond
          (str/starts-with? line "BookmarkTitle: ")
          (recur (rest lines)
                 (subs line (count "BookmarkTitle: "))
                 bookmarks)

          (and current-title (str/starts-with? line "BookmarkPageNumber: "))
          (let [page-text (subs line (count "BookmarkPageNumber: "))
                page-number (parse-int page-text "bookmark page number")]
            (recur (rest lines)
                   nil
                   (conj bookmarks {:title current-title :page page-number})))

          :else
          (recur (rest lines) current-title bookmarks))))))

(defn bookmark-page [bookmarks title-prefix]
  (if-let [match (first (filter #(str/starts-with? (:title %) title-prefix) bookmarks))]
    (:page match)
    (die (str "could not find bookmark starting with: " title-prefix))))

(defn range-start [ranges range-name]
  (:start (first (filter #(= (:name %) range-name) ranges))))

(defn build-physical-ranges [bookmarks]
  (let [chapter1-start (bookmark-page bookmarks "Chapter 1. ")
        appendix-start (bookmark-page bookmarks "Appendix A:")
        first-printed-page (:start (first printed-ranges))
        page-offset (- chapter1-start first-printed-page)]
    (when (neg? page-offset)
      (die (str "calculated negative page offset: " page-offset)))
    (let [front-matter-end (dec chapter1-start)]
      (when (< front-matter-end 1)
        (die (str "invalid front matter end page: " front-matter-end)))
      (let [physical-ranges (into [{:name "00-front-matter" :start 1 :end front-matter-end}]
                                  (map (fn [entry]
                                         (-> entry
                                             (update :start + page-offset)
                                             (update :end + page-offset)))
                                       printed-ranges))
            appendices-start (range-start physical-ranges "15-appendices")]
        (when (not= appendices-start appendix-start)
          (die (str "appendix start mismatch: range starts at "
                    appendices-start
                    ", bookmark starts at "
                    appendix-start)))
        {:ranges physical-ranges :page-offset page-offset}))))

(defn validate-ranges [ranges total-pages]
  (loop [remaining ranges
         previous-end 0
         line-number 1]
    (if (empty? remaining)
      (when (< previous-end total-pages)
        (warn (str "ranges end at "
                   previous-end
                   " but source has "
                   total-pages
                   " pages (trailing pages excluded)")))
      (let [{:keys [name start end]} (first remaining)]
        (when (> start end)
          (die (str "line " line-number " has start > end: " name " " start "-" end)))
        (when (<= start previous-end)
          (die (str "line "
                    line-number
                    " overlaps prior range: "
                    name
                    " starts at "
                    start
                    ", previous ended at "
                    previous-end)))
        (when (not= start (inc previous-end))
          (die (str "line "
                    line-number
                    " has gap before "
                    name
                    ": expected "
                    (inc previous-end)
                    ", got "
                    start)))
        (when (> end total-pages)
          (die (str "line "
                    line-number
                    " exceeds source page count ("
                    total-pages
                    "): "
                    name
                    " ends at "
                    end)))
        (recur (rest remaining) end (inc line-number))))))

(defn validate-bookmark-alignment [ranges bookmarks]
  (doseq [{:keys [range-name title-prefix]} bookmark-checks]
    (let [expected (bookmark-page bookmarks title-prefix)
          actual (range-start ranges range-name)]
      (when (nil? actual)
        (die (str "could not find generated range: " range-name)))
      (when (not= actual expected)
        (die (str "bookmark mismatch for "
                  range-name
                  ": range starts at "
                  actual
                  ", bookmark starts at "
                  expected))))))

(defn human-size [size-bytes]
  (let [units ["B" "K" "M" "G" "T"]]
    (loop [size (double size-bytes) index 0]
      (if (and (>= size 1024.0) (< index (dec (count units))))
        (recur (/ size 1024.0) (inc index))
        (if (zero? index)
          (format "%d%s" (long size) (nth units index))
          (format "%.1f%s" size (nth units index)))))))

(defn split-range [src-path out-dir {:keys [name start end]}]
  (let [pages (inc (- end start))
        out-path (.getPath (io/file out-dir (str name ".pdf")))]
    (println (format "  %s (%dpp: %d-%d)" name pages start end))
    (run-command "qpdf"
                 src-path
                 "--pages"
                 "."
                 (str start "-" end)
                 "--"
                 out-path)))

(defn list-output-pdfs [out-dir]
  (let [files (->> (.listFiles (io/file out-dir))
                   (filter #(str/ends-with? (.getName %) ".pdf"))
                   (sort-by #(.length %) >))]
    (doseq [file files]
      (println (format "%5s %s" (human-size (.length file)) (.getName file))))))

(defn default-base-dir []
  (let [cwd (System/getProperty "user.dir")
        candidates [cwd (str (io/file cwd "rp2350-reference"))]]
    (or (first (filter #(-> (io/file % "rp2350-datasheet.pdf") .exists) candidates))
        cwd)))

(defn total-pages [src-path]
  (parse-int (run-capture "qpdf" "--show-npages" src-path) "total page count"))

(defn -main [& args]
  (when (> (count args) 2)
    (binding [*out* *err*]
      (println "usage: clojure split-datasheet.clj [source.pdf] [output_dir]"))
    (System/exit 2))

  (let [base-dir (default-base-dir)
        src-path (if (>= (count args) 1)
                   (nth args 0)
                   (.getPath (io/file base-dir "rp2350-datasheet.pdf")))
        out-dir (if (>= (count args) 2)
                  (nth args 1)
                  (.getPath (io/file base-dir "datasheet")))]
    (require-tool "qpdf")
    (require-tool "pdftk")
    (when-not (.exists (io/file src-path))
      (die (str "source PDF not found: " src-path)))

    (let [bookmarks (parse-bookmarks (run-capture "pdftk" src-path "dump_data"))
          {:keys [ranges page-offset]} (build-physical-ranges bookmarks)
          pages (total-pages src-path)]
      (validate-ranges ranges pages)
      (validate-bookmark-alignment ranges bookmarks)

      (.mkdirs (io/file out-dir))
      (println (str "Splitting RP2350 datasheet from: " src-path " (page offset: +" page-offset ")"))
      (println)
      (doseq [range ranges]
        (split-range src-path out-dir range))
      (println)
      (println (str "Done! Files in: " out-dir))
      (list-output-pdfs out-dir))))

(apply -main *command-line-args*)

File: rp2350-reference/split-datasheet.py
#!/usr/bin/env python3
"""Split RP2350 datasheet into chapter PDFs using qpdf.

Usage:
    ./split-datasheet.py [source.pdf] [output_dir]

Notes:
    - Preferred implementation for readability.
    - Bash fallback remains available as ./split-datasheet.sh.
"""

from __future__ import annotations

import re
import shutil
import subprocess
import sys
from pathlib import Path
from typing import NoReturn


# Printed page ranges from datasheet footers. Converted to physical PDF pages
# using an automatically detected offset from chapter bookmarks.
PRINTED_RANGES: list[tuple[str, int, int]] = [
    ("01-introduction", 13, 23),
    ("02-system-bus", 24, 34),
    ("03a-processor-SIO", 35, 81),
    ("03b-processor-interrupts-debug", 82, 99),
    ("03c-cortex-m33-coprocessors", 100, 122),
    ("03d-cortex-m33-processor", 123, 232),
    ("03e-hazard3-processor", 233, 334),
    ("03f-architecture-switching", 335, 336),
    ("04-memory", 337, 352),
    ("05a-bootrom-concepts", 353, 374),
    ("05b-bootrom-apis", 375, 415),
    ("05c-bootrom-usb-uart", 416, 440),
    ("06-power", 441, 493),
    ("07-resets", 494, 512),
    ("08a-clocks-overview", 513, 553),
    ("08b-clocks-oscillators", 554, 586),
    ("09a-gpio-overview", 587, 603),
    ("09b-gpio-io-user-bank", 604, 759),
    ("09c-gpio-io-qspi-pads", 760, 815),
    ("10-security", 816, 875),
    ("11a-pio-overview-model", 876, 888),
    ("11b-pio-instructions", 889, 901),
    ("11c-pio-details-examples", 902, 960),
    ("12a-uart", 961, 982),
    ("12b-i2c", 983, 1045),
    ("12c-spi", 1046, 1065),
    ("12d-adc-temp", 1066, 1075),
    ("12e-pwm", 1076, 1093),
    ("12f-dma", 1094, 1140),
    ("12g-usb", 1141, 1181),
    ("12h-timers-watchdog", 1182, 1201),
    ("12i-hstx", 1202, 1211),
    ("12j-trng-sha256", 1212, 1225),
    ("12k-qspi-qmi", 1226, 1248),
    ("12l-system-control", 1249, 1267),
    ("13-otp", 1268, 1326),
    ("14-electrical", 1327, 1348),
    ("15-appendices", 1349, 1378),
]

BOOKMARK_ALIGNMENT: list[tuple[str, str]] = [
    ("01-introduction", r"^Chapter 1[.] "),
    ("02-system-bus", r"^Chapter 2[.] "),
    ("03a-processor-SIO", r"^Chapter 3[.] "),
    ("04-memory", r"^Chapter 4[.] "),
    ("05a-bootrom-concepts", r"^Chapter 5[.] "),
    ("06-power", r"^Chapter 6[.] "),
    ("07-resets", r"^Chapter 7[.] "),
    ("08a-clocks-overview", r"^Chapter 8[.] "),
    ("09a-gpio-overview", r"^Chapter 9[.] "),
    ("10-security", r"^Chapter 10[.] "),
    ("11a-pio-overview-model", r"^Chapter 11[.] "),
    ("12a-uart", r"^Chapter 12[.] "),
    ("13-otp", r"^Chapter 13[.] "),
    ("14-electrical", r"^Chapter 14[.] "),
    ("15-appendices", r"^Appendix A:"),
]


def die(message: str) -> NoReturn:
    print(f"error: {message}", file=sys.stderr)
    raise SystemExit(1)


def warn(message: str) -> None:
    print(f"warning: {message}", file=sys.stderr)


def run_capture(args: list[str]) -> str:
    try:
        result = subprocess.run(args, check=True, text=True, capture_output=True)
        return result.stdout
    except subprocess.CalledProcessError as exc:
        details = exc.stderr.strip() or exc.stdout.strip()
        die(f"command failed: {' '.join(args)} ({details or f'exit {exc.returncode}'})")


def run(args: list[str]) -> None:
    try:
        subprocess.run(args, check=True)
    except subprocess.CalledProcessError as exc:
        die(f"command failed: {' '.join(args)} (exit {exc.returncode})")


def require_prereqs(src: Path) -> None:
    for tool in ("qpdf", "pdftk"):
        if shutil.which(tool) is None:
            die(f"{tool} is required but not installed")
    if not src.is_file():
        die(f"source PDF not found: {src}")


def parse_bookmarks(bookmark_dump: str) -> list[tuple[str, int]]:
    bookmarks: list[tuple[str, int]] = []
    title: str | None = None

    for raw_line in bookmark_dump.splitlines():
        line = raw_line.strip()
        if line.startswith("BookmarkTitle: "):
            title = line[len("BookmarkTitle: ") :]
            continue
        if line.startswith("BookmarkPageNumber: ") and title is not None:
            page_text = line[len("BookmarkPageNumber: ") :]
            try:
                page_num = int(page_text)
            except ValueError:
                die(f"invalid bookmark page number: {page_text}")
            bookmarks.append((title, page_num))
            title = None

    if not bookmarks:
        die("no bookmarks found in PDF metadata")
    return bookmarks


def bookmark_page(bookmarks: list[tuple[str, int]], title_regex: str) -> int:
    pattern = re.compile(title_regex)
    for title, page_num in bookmarks:
        if pattern.search(title):
            return page_num
    die(f"could not find bookmark matching regex: {title_regex}")


def build_physical_ranges(
    bookmarks: list[tuple[str, int]],
) -> tuple[list[tuple[str, int, int]], int]:
    chapter1_start = bookmark_page(bookmarks, r"^Chapter 1[.] ")
    appendix_start = bookmark_page(bookmarks, r"^Appendix A: ")

    first_printed_page = PRINTED_RANGES[0][1]
    page_offset = chapter1_start - first_printed_page
    if page_offset < 0:
        die(f"calculated negative page offset: {page_offset}")

    front_end = chapter1_start - 1
    if front_end < 1:
        die(f"invalid front matter end page: {front_end}")

    physical_ranges: list[tuple[str, int, int]] = [("00-front-matter", 1, front_end)]
    for name, start, end in PRINTED_RANGES:
        physical_ranges.append((name, start + page_offset, end + page_offset))

    range_map = {name: (start, end) for name, start, end in physical_ranges}
    appendices_start = range_map["15-appendices"][0]
    if appendices_start != appendix_start:
        die(
            "appendix start mismatch: "
            f"range starts at {appendices_start}, bookmark starts at {appendix_start}"
        )

    return physical_ranges, page_offset


def validate_ranges(ranges: list[tuple[str, int, int]], total_pages: int) -> None:
    prev_end = 0
    for idx, (name, start, end) in enumerate(ranges, start=1):
        if start > end:
            die(f"line {idx} has start > end: {name} {start}-{end}")
        if start <= prev_end:
            die(
                "line "
                f"{idx} overlaps prior range: {name} starts at {start}, previous ended at {prev_end}"
            )
        if start != prev_end + 1:
            die(
                "line "
                f"{idx} has gap before {name}: expected {prev_end + 1}, got {start}"
            )
        if end > total_pages:
            die(
                f"line {idx} exceeds source page count ({total_pages}): "
                f"{name} ends at {end}"
            )
        prev_end = end

    if prev_end < total_pages:
        warn(
            f"ranges end at {prev_end} but source has {total_pages} pages "
            "(trailing pages excluded)"
        )


def validate_bookmark_alignment(
    ranges: list[tuple[str, int, int]],
    bookmarks: list[tuple[str, int]],
) -> None:
    starts = {name: start for name, start, _ in ranges}
    for name, title_regex in BOOKMARK_ALIGNMENT:
        expected = bookmark_page(bookmarks, title_regex)
        actual = starts.get(name)
        if actual is None:
            die(f"could not find generated range: {name}")
        if actual != expected:
            die(
                f"bookmark mismatch for {name}: "
                f"range starts at {actual}, bookmark starts at {expected}"
            )


def human_size(size_bytes: int) -> str:
    units = ["B", "K", "M", "G", "T"]
    size = float(size_bytes)
    unit_idx = 0
    while size >= 1024 and unit_idx < len(units) - 1:
        size /= 1024
        unit_idx += 1
    if unit_idx == 0:
        return f"{int(size)}{units[unit_idx]}"
    return f"{size:.1f}{units[unit_idx]}"


def split_range(src: Path, out_dir: Path, name: str, start: int, end: int) -> None:
    pages = end - start + 1
    print(f"  {name} ({pages}pp: {start}-{end})")
    out_pdf = out_dir / f"{name}.pdf"
    run(
        [
            "qpdf",
            str(src),
            "--pages",
            ".",
            f"{start}-{end}",
            "--",
            str(out_pdf),
        ]
    )


def main(argv: list[str]) -> int:
    script_dir = Path(__file__).resolve().parent
    src = Path(argv[1]).resolve() if len(argv) > 1 else script_dir / "rp2350-datasheet.pdf"
    out_dir = Path(argv[2]).resolve() if len(argv) > 2 else script_dir / "datasheet"
    if len(argv) > 3:
        print(f"usage: {Path(argv[0]).name} [source.pdf] [output_dir]", file=sys.stderr)
        return 2

    require_prereqs(src)
    bookmarks = parse_bookmarks(run_capture(["pdftk", str(src), "dump_data"]))
    ranges, page_offset = build_physical_ranges(bookmarks)

    total_pages_text = run_capture(["qpdf", "--show-npages", str(src)]).strip()
    try:
        total_pages = int(total_pages_text)
    except ValueError:
        die(f"invalid total page count from qpdf: {total_pages_text}")

    validate_ranges(ranges, total_pages)
    validate_bookmark_alignment(ranges, bookmarks)

    out_dir.mkdir(parents=True, exist_ok=True)
    print(f"Splitting RP2350 datasheet from: {src} (page offset: +{page_offset})")
    print()

    for name, start, end in ranges:
        split_range(src, out_dir, name, start, end)

    print()
    print(f"Done! Files in: {out_dir}")
    pdfs = sorted(out_dir.glob("*.pdf"), key=lambda path: path.stat().st_size, reverse=True)
    for pdf_path in pdfs:
        print(f"{human_size(pdf_path.stat().st_size):>5} {pdf_path.name}")

    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))
